### PR TITLE
Publish AS::Executor and AS::Reloader APIs

### PR DESCRIPTION
**Pull request by matthewd**
These should allow external code to run blocks of user code to do "work", at a similar unit size to a web request, without needing to get intimate with ActionDispatch.

The theory is that interested callers (Sidekiq, ActionCable) can just do:

``` ruby
Rails.application.reloader.wrap do
  # run some user code
end
```

.. and we'll take care of the interlock, code reloading, returning AR connections to the pool, and anything else that might be relevant.

### CODE REVIEW

**Comment by vipulnsward** [comment]
😱💚

**Comment by rafaelfranca** [comment]
This is really promising. I like the API and the implementation is simple. :+1:

File: railties/lib/rails/generators/rails/app/templates/config.ru [hunk 1]
```diff
@@ -1,10 +1,4 @@
 # This file is used by Rack-based servers to start the application.
 
 require ::File.expand_path('../config/environment', __FILE__)
-<%- unless options[:skip_action_cable] -%>
-
-# Action Cable requires that all classes are loaded in advance
-Rails.application.eager_load!
```
Comment (rafaelfranca) [inline railties/lib/rails/generators/rails/app/templates/config.ru]: Is this not needed anymore?

Comment (rafaelfranca) [inline railties/lib/rails/generators/rails/app/templates/config.ru]: Ah. it was because Action Cable, yeah, it is not needed.

**Comment by kaspth** [comment]
Love this :heart:

**Comment by jeremy** [comment]
🤘

**Comment by mikhailov** [comment]
:+1: Since we have an option to reload user code what you think to enhance the functionality to avoid code reload for static assets (idea came from https://github.com/rails/rails/pull/23518)

**Comment by spastorino** [comment]
The integration.rb part of the changes introduces an issue when using capybara with a driver that needs to start a server in a separate thread like poltergeist, selenium, etc.

Both threads, the test thread and the capybara server one run synchronize over the interlock so the server can't execute when receives the request because it's locked.

I don't know what was the intention of the changes in integration.rb were but doesn't seem to have sense, given that all it adds is this lock and QueryCache which doesn't make much sense per integration test as it's working with this changes.

Fixed here https://github.com/rails/rails/commit/cf075d9f55047281a5912e1b28cc4cb4bfcfdbf0

**Comment by mperham** [comment]
Would there be any benefit to using `call` instead of `wrap` as the method name?  More idiomatic or reusable?

**Comment by arthurnn** [comment]
I this changed the QueryCache to be an executor instead of a middleware. 
Before, to disable QueryCache in a rails app, I could just remove the middleware. (`config.middleware.delete "ActiveRecord::QueryCache"`). How can I disable it now? should we add some way to?

**Comment by mrrusof** [comment]
I see you touched `ActiveRecord::ConnectionAdapters::ConnectionManagement`. I'm in the process of figuring out how to reconcile Sinatra+Thin+ActiveRecord. The problem is that `ConnectionManagement` in ActiveRecord 4.2.6 does not return connections to the pool when used by Thin 1.6.4. `ConnectionManagement` returns to the pool those connections that are active and belong to the current thread. For each request, Thin connects to the database in a thread and applies `ConnectionManagement` in another.

Does this pull request resolve the problem? If not, what is "the right way" to approach the problem?

Related questions in stackoverflow:

- [ActiveRecord::ConnectionTimeoutError](http://stackoverflow.com/questions/13675879/activerecordconnectiontimeouterror)
- [ActiveRecord connection warning. (Database connections will not be closed automatically)](http://stackoverflow.com/questions/10191531/activerecord-connection-warning-database-connections-will-not-be-closed-automa)
- [How do I reuse connections in ActiveRecord?](http://stackoverflow.com/questions/36697685/how-do-i-reuse-connections-in-activerecord/)

**Comment by arthurnn** [comment]
> Does this pull request resolve the problem? If not, what is "the right way" to approach the problem?

@mrrusof I dont think it does. There is #24608 for instance to deal with connection pool, etc. if thats what you are asking.

**Comment by mrrusof** [comment]
@arthurnn Well, I created [an issue in Thin](https://github.com/macournoyer/thin/issues/307) and the author resolved the problem by finishing requests in the same thread that starts them.

**Pull request merged**